server/internal/infrastructure/aws/file.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (f *fileRepo) GetAssetFiles(ctx context.Context, u string) ([]gateway.FileEntry, error) {
+	p := getS3ObjectPath(u, "")
+	var fileEntries []gateway.FileEntry
+	err := f.s3Client.ListObjectsV2PagesWithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(f.bucketName),
+		Prefix: aws.String(p),
+	}, func(page *s3.ListObjectsV2Output, lastPage bool) bool {
+		for _, obj := range page.Contents {
+			fe := gateway.FileEntry{
+				Name: strings.TrimPrefix(*obj.Key, p),
+				Size: *obj.Size,
+			}
+			fileEntries = append(fileEntries, fe)
+		}
+		return true
+	})
+	if err != nil {
+		return nil, rerror.ErrInternalBy(err)
+	}
+	if len(fileEntries) == 0 {
+		return nil, gateway.ErrFileNotFound
+	}
+	return fileEntries, nil
+}
+
+func (f *fileRepo) UploadAsset(ctx context.Context, file *file.File) (string, int64, error) {
+	if file == nil {
+		return "", 0, gateway.ErrInvalidFile
+	}
+	if file.Size >= fileSizeLimit {
+		return "", 0, gateway.ErrFileTooLarge
+	}
+
+	fileUUID := newUUID()
+
+	p := getS3ObjectPath(fileUUID, file.Name)
+	if p == "" {
+		return "", 0, gateway.ErrInvalidFile
+	}
+
+	size, err := f.upload(ctx, p, file.Content)
+	if err != nil {
+		return "", 0, err
+	}
+	return fileUUID, size, nil
+}
+
+func (f *fileRepo) DeleteAsset(ctx context.Context, u string, fn string) error {
+	p := getS3ObjectPath(u, fn)
+	if p == "" {
+		return gateway.ErrInvalidFile
+	}
+	return f.delete(ctx, p)
+}
+
+func (f *fileRepo) GetURL(a *asset.Asset) string {
+	return getURL(f.baseURL.String(), a.UUID(), a.FileName())
+}
+
+func (f *fileRepo) IssueUploadAssetLink(ctx context.Context, filename, contentType string, expiresAt time.Time) (string, string, error) {
+	uuid := newUUID()
+
+	p := getS3ObjectPath(uuid, filename)
+	if p == "" {
+		return "", "", gateway.ErrInvalidFile
+	}
+
+	req, _ := f.s3Client.PutObjectRequest(&s3.PutObjectInput{
+		Bucket:       aws.String(f.bucketName),
+		CacheControl: aws.String(f.cacheControl),
+		Key:          aws.String(p),
+		ContentType:  aws.String(contentType),
+	})
+
+	uploadURL, err := req.Presign(15 * time.Minute)
+	if err != nil {
+		return "", "", err
+	}
+	return uploadURL, uuid, nil
+}
+
+func (f *fileRepo) UploadedAsset(ctx context.Context, u *asset.Upload) (*file.File, error) {
+	p := getS3ObjectPath(u.UUID(), u.FileName())
+
+	obj, err := f.s3Client.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(f.bucketName),
+		Key:    aws.String(p),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &file.File{
+		Content:     nil,
+		Name:        u.FileName(),
+		Size:        *obj.ContentLength,
+		ContentType: *obj.ContentType,
+	}, nil
+}
+
+func (f *fileRepo) read(ctx context.Context, filename string) (io.ReadCloser, error) {
+	if filename == "" {
+		return nil, rerror.ErrNotFound
+	}
+
+	buf := aws.NewWriteAtBuffer([]byte{})
+	_, err := f.s3Downloader.DownloadWithContext(ctx, buf, &s3.GetObjectInput{
+		Bucket: aws.String(f.bucketName),
+		Key:    aws.String(filename),
+	})
```

Comment:
The current implementation keeps all contents in memory. This means that simply trying to download a large file can easily run out of memory.

Instead of keeping it in memory, concatenate the streams and pass the data from S3 to the response.

```go
bucket := "your-bucket-name" 
key    := "path/to/key/filename.ext" 

resp, err := f.s3Client.GetObject(&s3.GetObjectInput{
    Bucket: aws.String(bucket),
    Key:    aws.String(key),
})
if err != nil {
	return rerror.ErrInternalBy(fmt.Errorf("s3: failed to get object: %w", err))
}

defer func() {
  _ = resp.Body.Close()
}()
return c.Stream(http.StatusOK, contentType, resp.Body)
```

Reply:
Done!